Implement clone functionality for widgets

samshara opened the pull request:
- fixes #1810
- fixes #2167

## Changes

* Add clone function for each widget type
## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] build works
- [x] eslint issues
- [x] typescript issues
- [x] `console.log` meant for debugging
- [x] typos
- [x] unwanted comments
- [x] conflict markers

## This PR contains valid:

- [x] permission checks
- [x] translations

samshara committed:
Implement clone functionality for widgets

tnagorra commented on app/views/AnalyticalFramework/utils.ts:
```diff
@@ -0,0 +1,183 @@
+    return ({
+        ...widget,
+        title: `${widget.title}-cloned`,
+        key: randomString(),
+        clientId: randomString(),
```
These two may need to have the same value.
@thenav56 what do you think?

thenav56 replied:
Are we now using key instead of clientId? Or both?

Not sure where this is used. But if we have both, then we can easily remove clientId in the future (Currently server uses clientId for widget properties data)

tnagorra replied:
I have started work on PR to convert all "persistent" keys to be named "key".

One problem is that we have "id", "clientId" and "key" on widget level.
We could do away with "key", but don't know how its used on the server

tnagorra commented on app/views/AnalyticalFramework/utils.ts:
```diff
@@ -0,0 +1,183 @@
+function clone(widget: PartialWidget) {
+    return ({
+        ...widget,
+        title: `${widget.title}-cloned`,
+        key: randomString(),
+        clientId: randomString(),
+        id: undefined,
```
We should be able to re-use this logic for other widgets a well.
Something like the example below:

```javascript
// NOTE: we need to extend only the basic properties that we are modifying

function cloneWidgetSuperficially<T extends BaseWidget>(value: T): T {
    const randKey = randomString();
    return {
        ...value,
        key: randKey,
        clientId: randKey,
        id: undefined,
        title: `${value.title}: Cloned`
    };
}
```

tnagorra commented on app/views/AnalyticalFramework/utils.ts:
```diff
@@ -0,0 +1,183 @@
+        case 'TEXT':
+            return clone(widget);
+        default:
+            return undefined;
```
Haven't we used this utiltiy on primary tagging and secondary tagging page?

samshara committed:
Reuse functions to clone widgets and options

samshara committed:
Implement widget clone for all widgets in AnalyticalFramework

tnagorra committed:
Simplify clone logic

- Add widget size input in organigram widget edit
- Use same trash icon everywhere in deep

AdityaKhatri committed:
Fix clone widget typings
